Integrated all subchapters of Module 2 into Quarto website (#186)
UBC-MDS/exploratory-data-viz: modules/module2/module2-08-true_or_false_facets.qmd

---
format:
  live-html
resources:
  - ../utils.py
  - ../data
---

<script src='../../src/quiz.js'></script>  

# 4.1. Exercises

## True or False: Facets

<div id='mcq1'></div> 
<script>
    generateQuiz(
        'mcq1', 
        'True or False',  

>         'If you faceting and assign a colour channel to the same column, you will obtain subplots all in different colours.',  

Reviewer: ```suggestion
        'If you facet and assign a colour channel to the same column, each subplot will contain only one colour.',  
```

        { 
        'True': '',

>         'False': 'Setting a colour channel to the same column as the facet will change each of the faceted plots to different colours.'

Reviewer: ```suggestion
        'False': 'Since there is only one group (based on the chosen column) in each faceted chart, there will also only be one color in each of the charts.'
```

- Make sure to give it a title and set the dimensions to a height of 100 and a width of 150.
- Facet your plots by `species` for the columns and by `island` for the rows.
- Save your plot in an object named `penguin_facet`.

```{pyodide}
#| setup: true
#| exercise: faceting_with_penguins

import altair as alt
import pandas as pd
from utils import assert_chart_equal 

penguins_df = pd.read_csv('data/penguins.csv')
```

```{pyodide}
#| exercise: faceting_with_penguins
penguin_facet = alt.Chart(____).____().encode(
    x=____('body_mass_g', ____),
    y=____,
    ).properties(____='Penguin mass', ____, ____
    ).____(____=____, ____='island')

penguin_facet
```

```{pyodide}
#| exercise: faceting_with_penguins
#| check: true

solution = alt.Chart(penguins_df).mark_bar().encode(
    x=alt.X('body_mass_g', bin=alt.Bin(maxbins=15)),
    y='count()',
    ).properties(title='Penguin mass', width=150, height=100
    ).facet(column='species', row='island')

assert_chart_equal(solution, result)
```

:::: { .hint exercise="faceting_with_penguins"}
::: { .callout-note collapse="false"}

## Hint 1

- Are you using `.mark_bar()`?
- Are you assigning the plot to an object named`penguin_facet`?

> - Are you specifying `alt.X('species', bin=True)` on the x-axis?

Reviewer: ```suggestion
- Are you specifying `alt.X('species', bin=alt.Bin(maxbins=15))` on the x-axis?
```

- Are you specifying `count()` on the y-axis?
- Are you giving the plot a title and proper dimensions using `.properties()`?

:::
::::

:::: { .solution exercise="faceting_with_penguins" }
::: { .callout-tip collapse="false"}

## Fully worked solution:

```{pyodide}
penguin_facet = alt.Chart(penguins_df).mark_bar().encode(
    x=alt.X('body_mass_g', bin=alt.Bin(maxbins=15)),
    y='count()',

>     ).properties(title='Penguin mass', width=150, height=100
>     ).facet(column='species', row='island')

Reviewer: ```suggestion
).properties(title='Penguin mass', width=150, height=100
).facet(column='species', row='island')
```

import pandas as pd
from utils import assert_chart_equal 

def remove_keys_inplace(spec, keys):
    if isinstance(spec, dict):
        for key in list(spec.keys()):
            if key in keys:
                del spec[key]
            else:
                remove_keys_inplace(spec[key], keys)
    elif isinstance(spec, list):
        for item in spec:

>             remove_keys_inplace(item, keys)

Reviewer: If you are using this in multiple places, could we move it to `utils.py` and just import it here?
